crates/cli/src/detect.rs

```diff
@@ -0,0 +1,68 @@
+use duct::cmd;
+use std::io;
+use std::path::Path;
+
+/// Find an executable in the `PATH`.
+pub(crate) fn find_executable(exe_name: impl AsRef<Path>) -> Option<std::path::PathBuf> {
+    std::env::var_os("PATH").and_then(|paths| {
+        std::env::split_paths(&paths)
+            .filter_map(|dir| {
+                let full_path = dir.join(&exe_name);
+                full_path.is_file().then_some(full_path)
+            })
+            .next()
+    })
+}
+
+/// Check if `rustup` is installed (aka: Is in the `PATH`).
+pub(crate) fn has_rust_up() -> anyhow::Result<bool> {
+    Ok(match std::env::consts::OS {
+        "linux" | "freebsd" | "netbsd" | "openbsd" | "solaris" | "macos" => find_executable("rustup").is_some(),
+        "windows" => find_executable("rustup.exe").is_some(),
+        unsupported_os => {
+            return Err(anyhow::anyhow!(
+                "This OS may be unsupported for `rustup`: {unsupported_os}"
+            ));
+        }
+    })
+}
+
+/// Check if `rustfmt` is installed (aka: Is in the `PATH`).
+pub(crate) fn has_rust_fmt() -> anyhow::Result<bool> {
+    Ok(match std::env::consts::OS {
+        "linux" | "freebsd" | "netbsd" | "openbsd" | "solaris" | "macos" => find_executable("rustfmt").is_some(),
+        "windows" => find_executable("rustfmt.exe").is_some(),
+        unsupported_os => {
+            return Err(anyhow::anyhow!(
```

Comment:
I think we'd want to just return `false` here - if OS is unsupported by Rustup, that doesn't mean it's not supported by Rust altogether, so no need to return an error.